pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-schema-objects

```diff
@@ -0,0 +1,182 @@
+new-server name=s1 nodes=1 splits=1000
+----
+
+subtest restore-cleanup
+
+# Disable GC job so that the final check of crdb_internal.tables is
+# guaranteed to not be cleaned up. Although this was never observed by a
+# stress test, it is here for safety.
+exec-sql
+SET CLUSTER SETTING jobs.debug.pausepoints = 'gcjob.before_resume';
+----
+
+# Create a user defined type and check that it is cleaned up after the
+# failed restore.
+exec-sql
+CREATE DATABASE restore;
+CREATE TYPE data.myenum AS ENUM ('hello');
+----
+
+# Do the same with a user defined schema.
+exec-sql
+USE data;
+CREATE SCHEMA myschema;
+----
+
+exec-sql
+BACKUP DATABASE data INTO 'nodelocal://1/foo';
+----
+
+exec-sql
+SET CLUSTER SETTING jobs.debug.pausepoints = 'restore.after_publishing_descriptors';
+----
+
+restore expect-pausepoint tag=a
+RESTORE data.* FROM LATEST IN 'nodelocal://1/foo' WITH into_db = 'restore';
+----
+job paused at pausepoint
+
+# Cancel the job so that the cleanup hook runs.
+job cancel=a
+----
+
+# Verify the cancelled RESTORE added some DROP tables.
+query-sql
+SELECT name FROM crdb_internal.tables WHERE database_name = 'restore' AND state = 'DROP'
+----
+bank
+
+# Verify that `myenum` was cleaned out from the failed restore. There should
+# only be one namespace entry (data.myenum).
+query-sql
+SELECT count(*) FROM system.namespace WHERE name = 'myenum';
+----
+1
+
+# Check the same for data.myschema.
+query-sql
+SELECT count(*) FROM system.namespace WHERE name = 'myschema';
+----
+1
+
+# Verify that the only schema that appears is the public schema.
+query-sql
+USE restore;
+SHOW SCHEMAS;
+----
+crdb_internal <nil>
+information_schema <nil>
+pg_catalog <nil>
+pg_extension <nil>
+public admin
+
+# Ensure that a database level restore also cleans up after itself.
+exec-sql
```

Comment:
add. "i.e. this drop would fail if descriptors for the failed restore into `data` were not in a dropped state"